internal/utils/archive_extract.go

```diff
@@ -0,0 +1,102 @@
+package utils
+
+import (
+	"archive/tar"
+	"compress/gzip"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/ossf/package-analysis/internal/log"
+)
+
+// ExtractTarGzFile extracts a .tar.gz / .tgz file located at path, using
+// outputDir as the root of the extracted files.
+func ExtractTarGzFile(path string, outputDir string) error {
+	return processGzipFile(path, func(reader io.Reader) error {
+		return extractTar(reader, outputDir)
+	})
+}
+
+func processGzipFile(path string, process func(io.Reader) error) error {
+	gzFile, err := os.Open(path)
+	if err != nil {
+		return err
+	}
+
+	var unzippedBytes *gzip.Reader
+	if unzippedBytes, err = gzip.NewReader(gzFile); err != nil {
+		return err
+	}
+
+	defer func() {
+		if closeErr := unzippedBytes.Close(); closeErr != nil {
+			log.Error("failed to close gzip reader", "error", closeErr)
+		}
+	}()
+
+	if err = process(unzippedBytes); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+/*
+extractTar extracts the contents of the given stream of bytes of a tar archive, using
+outputDir as the root of the extracted files.
+*/
+func extractTar(tarStream io.Reader, outputDir string) error {
+	if outputDir == "" {
+		return fmt.Errorf("outputDir is empty")
+	}
+
+	tarReader := tar.NewReader(tarStream)
+
+	var header *tar.Header
+	var err error
+	for header, err = tarReader.Next(); err == nil; header, err = tarReader.Next() {
+		outputPath := filepath.Join(outputDir, header.Name)
+		// check for ZipSlip: https://snyk.io/research/zip-slip-vulnerability
+		if !strings.HasPrefix(outputPath, filepath.Clean(outputDir)) {
```

Comment:
I'm not sure this works. 

If outputPath is "/outputPath/../foo", this passes the HasPrefix check right? 

I think this should be:

```golang
strings.HasPrefix(filepath.Clean(outputPath), filepath.Clean(outputDir)+string(os.PathSeparator)
```

`filePath.Clean` is needed to normalize both paths for comparison. e.g. "/outputPath/../foo" would turn into "/foo" before it's used for comparison. 

We should also add a os.Separator to the prefix to prevent "/outputPath/../outputPathFOO" from passing this check either.

Reply:
Looking again at https://github.com/ossf/scorecard/blob/469216299ac3e161ebb995302ff8eb4ac43bb199/clients/githubrepo/tarball.go#LL49C23-L49C23, it looks like it's possible the scorecard implementation is wrong too.

Reply:
Oh, not sure if you accounted for this but `filePath.Join()` also does a `Clean()` internally

Reply:
Ok so turns out the first case was fine because of the fact that `filePath.Join()` does a `Clean()`, but adding `os.PathSeparator` was needed to avoid the second edge case. I've added tests for both cases

Reply:
Ah yep, I misremembered the `filePath.Join()` behaviour. Sorry about that!